### PR TITLE
Removing creationTimestamp under Template object

### PR DESCRIPTION
**Pull request by dperaza4dustbit**
The creationTimestamp field not only pops up under TemplateTier metadata but also under template itself

### CODE REVIEW

**Review by alexeykazakov — COMMENTED** [review]
Looks good to me but it looks like there is a room for improvements in the unit tests.

**Comment by dperaza4dustbit** [comment]
Will try to figure out if I spot the issue with unit test

File: pkg/cmd/generate/util.go [hunk 1]
```diff
@@ -159,6 +159,7 @@ const header = `# --------------------------------------------------------------
 func writeFile(filePath string, content []byte) error {
 	// https://github.com/kubernetes/kubernetes/issues/67610
 	contentString := strings.ReplaceAll(string(content), "\n  creationTimestamp: null", "")
+	contentString = strings.ReplaceAll(contentString, "\n      creationTimestamp: null", "")
```
Comment (fbm3307) [inline pkg/cmd/generate/util.go:162]: Not sure why the previous  replace didnt work. 
if having one function at a time works, then probably we should split the functionality , have the string value  of content assigned first, and then do the replace.

**Review by fbm3307 — COMMENTED** [review]
Apart from splitting the functions, if possible as a part of this PR we can also have a unit test to check these, they are currently not being tested, i am unsure of the reason for it

File: pkg/cmd/generate/util.go [hunk 2]
```diff
@@ -159,6 +159,7 @@ const header = `# --------------------------------------------------------------
 func writeFile(filePath string, content []byte) error {
 	// https://github.com/kubernetes/kubernetes/issues/67610
 	contentString := strings.ReplaceAll(string(content), "\n  creationTimestamp: null", "")
+	contentString = strings.ReplaceAll(contentString, "\n      creationTimestamp: null", " {}")
```
Comment (dperaza4dustbit) [inline pkg/cmd/generate/util.go:162]: Need to set to empty object to fix issue with ArgoCD synch loop created by TierTemplates do to spec.template.matadata see the fix in infra deployment here: https://github.com/redhat-appstudio/infra-deployments/pull/4120. The requirement is that all TierTemplates in source will start with empty object {} and then we need to remove the creationTimeStap set to null. If there is a better way to do this let me know.

**Comment by dperaza4dustbit** [comment]
Working on fixing the testcases

File: pkg/cmd/generate/util.go [hunk 3]
```diff
@@ -159,6 +159,7 @@ const header = `# --------------------------------------------------------------
 func writeFile(filePath string, content []byte) error {
 	// https://github.com/kubernetes/kubernetes/issues/67610
 	contentString := strings.ReplaceAll(string(content), "\n  creationTimestamp: null", "")
```
Comment (fbm3307) [inline pkg/cmd/generate/util.go:161]:
```suggestion
	contentString := string(content)
```

File: pkg/cmd/generate/util.go (hunk 2 above)
Comment (fbm3307) [inline pkg/cmd/generate/util.go:162]:
```suggestion
	contentString := strings.ReplaceAll(string(content), "\n      creationTimestamp: null", " {}")
```

**Review by fbm3307 — COMMENTED** [review]
This looks good, 
just a minor suggestion

File: pkg/cmd/generate/util.go (hunk 1 above)
Comment (fbm3307) [reply pkg/cmd/generate/util.go]: please ignore the comments, i thinks both the functions are required

File: pkg/cmd/generate/util.go (hunk 3 above)
Comment (fbm3307) [reply pkg/cmd/generate/util.go:161]: please ignore the comments, i thinks both the functions are required

File: pkg/cmd/generate/util.go (hunk 2 above)
Comment (fbm3307) [reply pkg/cmd/generate/util.go:162]: please ignore the comments, i thinks both the functions are required

Comment (MatousJobanek) [reply pkg/cmd/generate/util.go:162]: let's fix this empty metadata as a separate PR as this won't work if the spec.template.metadata.name is set (for example in our e2e tests).
Let's focus on the empty creationTimestamp for now

Comment (MatousJobanek) [inline pkg/cmd/generate/util.go:162]: I would rather make it more generic so it replaces any empty creationTimestamp no matter at which level
```suggestion
	emptyCreationTimestamp := regexp.MustCompile(`\n *creationTimestamp: null`)
	contentString := emptyCreationTimestamp.ReplaceAllString(string(content), "")
```

**Comment by MatousJobanek** [comment]
as for unit-testing - let's either improve the util_test.go so it verifies objects at any level, or add this check:
https://github.com/kubesaw/ksctl/blob/d75821023d8faa2b97e2b01761f090c0d7c53551/pkg/cmd/generate/assertion_test.go#L169-L171 
into
https://github.com/kubesaw/ksctl/blob/d75821023d8faa2b97e2b01761f090c0d7c53551/pkg/cmd/generate/assertion_test.go#L142-L143
because the `assertObject` function is not called in the `nstemplatetiers_test.go` but `listObjects` is, so it would be verified automatically.

File: pkg/cmd/generate/util.go [hunk 4]
```diff
@@ -158,8 +159,11 @@ const header = `# --------------------------------------------------------------
 
 func writeFile(filePath string, content []byte) error {
 	// https://github.com/kubernetes/kubernetes/issues/67610
-	contentString := strings.ReplaceAll(string(content), "\n  creationTimestamp: null", "")
+	emptyCreationTimestamp := regexp.MustCompile(`\n *creationTimestamp: null`)
```
Comment (dperaza4dustbit) [inline pkg/cmd/generate/util.go:162]: This works well as suggest by @MatousJobanek

File: pkg/cmd/generate/util.go [hunk 5]
```diff
@@ -158,8 +159,11 @@ const header = `# --------------------------------------------------------------
...
+	contentString := emptyCreationTimestamp.ReplaceAllString(string(content), "")
 	contentString = strings.ReplaceAll(contentString, "\nuser: {}", "")
+	// This will only apply to konflux tier template that start with metadata = {}
+	contentString = strings.ReplaceAll(contentString, "metadata:\n    objects:", "metadata: {}\n    objects:")
```
Comment (dperaza4dustbit) [inline pkg/cmd/generate/util.go:166]: This fixes the metadata problem with Konflux as tiers in src already start with spec.template.metadata as {}. See here: https://github.com/redhat-appstudio/infra-deployments/blob/main/components/sandbox/tiers/src/appstudio/spacerole_admin.yaml#L3
But this won't affect our e2e nor unitest as those start with spec.template.metadata.name having a string value. Would this work?

File: pkg/cmd/generate/util.go (hunk 2 above)
Comment (dperaza4dustbit) [reply pkg/cmd/generate/util.go]: see my new approach in utils.go, seems simpler to me

Comment (dperaza4dustbit) [reply pkg/cmd/generate/util.go]: Took your suggestion here

**Review by dperaza4dustbit — COMMENTED** [review]
Some comments added, working on improving test to check all timestamps

File: pkg/cmd/generate/util.go (hunk 5 above)
Comment (MatousJobanek) [reply pkg/cmd/generate/util.go:166]: looks good, this should work fine :+1: :rocket:

**Review by MatousJobanek — COMMENTED** [review]
it looks good overall, just waiting for the updates of the unit tests https://github.com/kubesaw/ksctl/pull/50#issuecomment-2242812687

File: pkg/cmd/generate/util.go (hunk 4 above)
Comment (dperaza4dustbit) [reply pkg/cmd/generate/util.go:166]: Done

**Comment by dperaza4dustbit** [comment]
> it looks good overall, just waiting for the updates of the unit tests [#50 (comment)](https://github.com/kubesaw/ksctl/pull/50#issuecomment-2242812687)

Got it

**Review by MatousJobanek — APPROVED** [review]
Looks good, thanks for addressing our comments

**Pull request merged**